x-pack/plugins/endpoint/server/routes/alerts/list/lib.ts

```diff
@@ -105,41 +152,71 @@ export const buildAlertListESQuery = async (reqData: AlertRequestData): Promise<
+  endpointAppContext: EndpointAppContext,
+  reqData: AlertListRequestData,
+  searchResponse: SearchResponse<AlertData>
+): AlertResultList {
+  let totalNumberOfAlerts: number = 0;
+  let totalIsLowerBound: boolean = false;
+
+  // This case is due to: https://github.com/elastic/kibana/issues/56694
+  const total: ESTotal = (searchResponse?.hits?.total as unknown) as ESTotal;
+  totalNumberOfAlerts = total?.value || 0;
+  totalIsLowerBound = total?.relation === 'gte' || false;
+
+  if (totalIsLowerBound) {
+    // This shouldn't happen, as we always try to fetch enough hits to satisfy the current request and the next page.
+    endpointAppContext.logFactory
+      .get('alerts')
+      .warn('Total hits not counted accurately. Pagination numbers may be inaccurate.');
+  }
 
-  if (request?.route?.method === 'get') {
-    reqData.pageIndex = request.query?.page_index;
-    reqData.pageSize = request.query?.page_size || config.alertResultListDefaultPageSize;
-    reqData.filters = request.query?.filters || config.alertResultListDefaultFilters;
-    reqData.query = request.query?.query || config.alertResultListDefaultQuery;
-    reqData.sort = request.query?.sort || config.alertResultListDefaultSort;
-    reqData.order = request.query?.order || config.alertResultListDefaultOrder;
-    reqData.searchAfter = request.query?.after;
-    reqData.searchBefore = request.query?.before;
-  } else {
-    reqData.pageIndex = request.body?.page_index;
-    reqData.pageSize = request.body?.page_size || config.alertResultListDefaultPageSize;
-    reqData.filters = request.body?.filters || config.alertResultListDefaultFilters;
-    reqData.query = request.body?.query || config.alertResultListDefaultQuery;
-    reqData.sort = request.body?.sort || config.alertResultListDefaultSort;
-    reqData.order = request.body?.order || config.alertResultListDefaultOrder;
-    reqData.searchAfter = request.body?.after;
-    reqData.searchBefore = request.body?.before;
+  const hits: AlertHits = searchResponse?.hits?.hits;
+  const hitLen: number = hits.length;
+
+  if (reqData.searchBefore !== undefined) {
+    // Reverse the hits if we used `search_before`.
+    hits.reverse();
   }
 
-  if (reqData.searchAfter === undefined && reqData.searchBefore === undefined) {
-    // simple pagination
-    if (reqData.pageIndex === undefined) {
-      reqData.pageIndex = config.alertResultListDefaultFirstPageIndex;
-    }
+  const firstTimestamp: Date = hits[0]._source['@timestamp'];
+  const lastTimestamp: Date = hits[hitLen - 1]._source['@timestamp'];
 
-    reqData.fromIndex = reqData.pageIndex * reqData.pageSize;
+  const firstEventId: number = hits[0]._source.event.id;
+  const lastEventId: number = hits[hitLen - 1]._source.event.id;
+
+  let pageUrl: string = '/api/endpoint/alerts?';
+  pageUrl +=
+    'filters=' +
+    reqData.filters +
+    '&page_size=' +
+    reqData.pageSize +
+    '&sort=' +
+    reqData.sort +
+    '&order=' +
+    reqData.order;
```

Comment:
Will interpolation not suffice here? May make it more readable as a format. With formatted string like this, it also helps to add a comment showing an example of completed string.

Reply:
@nnamdifrankie Will change to use interpolation!